vsintegration/src/FSharp.Editor/Common/CodeAnalysisExtensions.fs

```diff
@@ -0,0 +1,74 @@
+
+open Microsoft.CodeAnalysis
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+
+
+type Project with
+
+    member this.GetDependentProjectIds () =
+        this.Solution.GetProjectDependencyGraph().GetProjectsThatDirectlyDependOnThisProject(this.Id)
+
+    /// The list all projects within the same solution that reference this project.
+    member this.GetDependentProjects () =
+        this.Solution.GetProjectDependencyGraph().GetProjectsThatDirectlyDependOnThisProject(this.Id)
+        |> Seq.map this.Solution.GetProject
+
+
+    /// The list the ProjectIds of all of the projects that this project directly or transitively depneds on
+    member this.GetProjectIdsOfAllProjectsThisProjectDependsOn () =
+        let graph = this.Solution.GetProjectDependencyGraph()
+        let transitiveDependencies = graph.GetProjectsThatThisProjectTransitivelyDependsOn this.Id
+        let directDependencies = graph.GetProjectsThatThisProjectDirectlyDependsOn this.Id
+        Seq.append directDependencies transitiveDependencies
+        
+    /// The list all of the projects that this project directly or transitively depneds on
+    member this.GetAllProjectsThisProjectDependsOn () =
+        this.GetProjectIdsOfAllProjectsThisProjectDependsOn ()
+        |> Seq.map this.Solution.GetProject
+
+type Solution with 
+
+    /// Try to get a document inside the solution using the document's name
+    member self.TryGetDocumentNamed docName =
+        self.Projects |> Seq.tryPick (fun proj ->
+            proj.Documents |> Seq.tryFind (fun doc -> doc.Name = docName))
+
+        
+    member self.TryGetDocumentFromPath filePath =
+        self.GetDocumentIdsWithFilePath filePath
+        |> Seq.tryHead |> Option.map (fun docId -> self.GetDocument docId)
+
+    /// Try to get a project inside the solution using the project's id
+    member self.TryGetProject (projId:ProjectId) =
+        if self.ContainsProject projId then Some (self.GetProject projId) else None
+
+
+    member self.GetDependentProjects (projectId:ProjectId) =
+        self.GetProject(projectId).GetDependentProjects()
+
+    member self.GetDependentProjectIds (projectId:ProjectId) =
+        self.GetProject(projectId).GetDependentProjectIds()
+
+    /// Try to find the document assoicated with the F# range in the solution
+    /// If provided a projectId prefer a copy of the document from that project
+    member self.TryGetDocumentFromFSharpRange (range:range,?projectId:ProjectId) =
+        /// Retrieve the DocumentId from the workspace for the range's file
+        let filePath = System.IO.Path.GetFullPathSafe range.FileName
+        let candidates = self.GetDocumentIdsWithFilePath filePath
+        candidates|> Seq.iter(fun c -> logInfof "%s" (self.GetProject c.ProjectId).FilePath)
+        match candidates.Length with 
+        | 0 -> None 
+        | 1 -> Some (self.GetDocument candidates.[0])
+        | _ -> 
```

Comment:
I think it's the same functionality  as [https://github.com/Microsoft/visualfsharp/blob/13abe6f3b9380e78a51c7a53a04aac27d02304b8/vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs#L62-L68](https://github.com/Microsoft/visualfsharp/blob/13abe6f3b9380e78a51c7a53a04aac27d02304b8/vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs#L62-L68)